tests/integration/v3_watch_restore_test.go

```diff
@@ -90,19 +94,38 @@ func TestV3WatchRestoreSnapshotUnsync(t *testing.T) {
 	kvc := integration.ToGRPC(clus.Client(1)).KV
 
 	// to trigger snapshot from the leader to the stopped follower
-	for i := 0; i < 15; i++ {
+	//
+	// NOTE: In 3 members cluster, after initial lead has been elected,
+	// there are 3xConfChange + 3xMemberAttrSet + 1xClusterVersionSet logs.
+	// So member 0 has index 8 in raft log before network partition. We
+	// need to trigger snapshot at least twice.
+	//
+	// SnapshotCount: 10, SnapshotCatchUpEntries: 5
+	// T1: L(snapshot-index: 11, compacted-index:  6), F_m0(index:8)
+	// T2: L(snapshot-index: 22, compacted-index: 17), F_m0(index:8, out of date)
+	// T3: L(snapshot-index: 33, compacted-index: 28), F_mo(index:8, out of date)
+	//
+	// Since the snapshot is handled in GoAttach, we need to trigger
```

Comment:
The key part for leader to send snapshot is to to get leader to [compact](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/server.go#L2114) index 8. If you change SnapshotCatchUpEntries to 1, this can be achieved in 1 call of EtcdServer.snapshot.

Reply:
Correct. It was my first patch. But I run into other error.

```diff
diff --git a/tests/integration/v3_watch_restore_test.go b/tests/integration/v3_watch_restore_test.go
index bdebeacfc..552f4aecf 100644
--- a/tests/integration/v3_watch_restore_test.go
+++ b/tests/integration/v3_watch_restore_test.go
@@ -57,7 +57,7 @@ func TestV3WatchRestoreSnapshotUnsync(t *testing.T) {
        clus := integration.NewCluster(t, &integration.ClusterConfig{
                Size:                   3,
                SnapshotCount:          10,
-               SnapshotCatchUpEntries: 5,
+               SnapshotCatchUpEntries: 1,
        })
        defer clus.Terminate(t)
```

```bash
➜  integration git:(main) ✗ taskset -c 0,1 go test -v -count=1000 -timeout=700m -failfast -run TestV3WatchRestoreSnapshotUnsync ./

... 
    logger.go:130: 2023-04-08T09:24:52.907+0800 INFO    m1      saved snapshot  {"member": "m1", "snapshot-index": 22}
    logger.go:130: 2023-04-08T09:24:52.907+0800 INFO    m1      compacted Raft logs     {"member": "m1", "compact-index": 21}
    logger.go:130: 2023-04-08T09:24:52.907+0800 INFO    m0.raft d6a69b975d0a7658 became follower at term 2      {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.907+0800 INFO    m0.raft raft.node: d6a69b975d0a7658 elected leader c0c978b536d663c2 at term 2   {"member": "m0"}
{"level":"warn","ts":"2023-04-08T09:24:52.90773+0800","logger":"etcd-client","caller":"v3/retry_interceptor.go:65","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc00233d860/localhost:m0","method":"/etcdserverpb.KV/Range","attempt":0,"error":"rpc error: code = Unavailable desc = etcdserver: leader changed"}
    logger.go:130: 2023-04-08T09:24:52.911+0800 INFO    m2      saved snapshot  {"member": "m2", "snapshot-index": 22}
    logger.go:130: 2023-04-08T09:24:52.911+0800 INFO    m2      skip compaction since there is an inflight snapshot     {"member": "m2"}
    logger.go:130: 2023-04-08T09:24:52.911+0800 INFO    m2      sending database snapshot       {"member": "m2", "snapshot-index": 22, "remote-peer-id": "d6a69b975d0a7658", "bytes": 33666, "size": "34 kB"}
    logger.go:130: 2023-04-08T09:24:52.911+0800 INFO    m2      sending merged snapshot {"member": "m2", "from": "c0c978b536d663c2", "to": "d6a69b975d0a7658", "bytes": 33666, "size": "34 kB"}
    logger.go:130: 2023-04-08T09:24:52.911+0800 INFO    m2      sent database snapshot to writer        {"member": "m2", "bytes": 24576, "size": "25 kB"}
    logger.go:130: 2023-04-08T09:24:52.911+0800 INFO    m0      receiving database snapshot     {"member": "m0", "local-member-id": "d6a69b975d0a7658", "remote-snapshot-sender-id": "c0c978b536d663c2", "incoming-snapshot-index": 22, "incoming-snapshot-message-size-bytes": 9090, "incoming-snapshot-message-size": "9.1 kB"}

...

    logger.go:130: 2023-04-08T09:24:52.915+0800 INFO    m0      restoring lease store   {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.917+0800 INFO    m0      restored lease store    {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.917+0800 INFO    m0      restoring mvcc store    {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.917+0800 INFO    m0      kvstore restored        {"member": "m0", "current-rev": 15}
    logger.go:130: 2023-04-08T09:24:52.917+0800 INFO    m0      restored mvcc store     {"member": "m0", "consistent-index": 22}
    logger.go:130: 2023-04-08T09:24:52.917+0800 INFO    m0      restoring alarm store   {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.917+0800 INFO    m0      closing old backend file        {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.918+0800 INFO    m0      restored alarm store    {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.918+0800 INFO    m0      restoring auth store    {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.918+0800 INFO    m0      restored auth store     {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.918+0800 INFO    m0      restoring v2 store      {"member": "m0"}
    logger.go:130: 2023-04-08T09:24:52.919+0800 INFO    m0      restored v2 store       {"member": "m0"}

...

    v3_watch_restore_test.go:128: sleeping for 2 seconds DONE
    v3_watch_restore_test.go:153: wStream.Recv error: expected 12 events, got [kv:<key:"foo" create_revision:2 mod_revision:5 version:4 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:6 version:5 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:7 version:6 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:8 version:7 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:9 version:8 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:10 version:9 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:11 version:10 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:12 version:11 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:13 version:12 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:14 version:13 value:"bar" >  kv:<key:"foo" create_revision:2 mod_revision:15 version:14 value:"bar" > ]
```

m0 should recover to have (consistent-index:23, current-rev:16). But it got (consisten-index:22, curent-rev:15). 
It is timing issue. 

https://github.com/etcd-io/etcd/blob/7153a8f2f48bc36dec1caafb97001634f0366fc4/server/etcdserver/server.go#L938

The server uses old index which doesn't match the snapshot content. Not sure that it is bug or fault-tolerant. Finally, it can catch-up. However, the watch event is not expected. cc @ahrtr 

So I try to make this patch to let it has time to handle memory compact. 
I am still testing it with limit core by `taskset -c`. Will update it later.
(If it doesn't work, I will switch to watch the log...)

Reply:
> However, the watch event is not expected. cc @ahrtr

Sorry, I do not get the point, and also did not get time dig into it for now. Please file a separate issue if you think it's an issue or something you can't explain or understand.